### PR TITLE
[Path] Deburr bug fixes

### PR DESCRIPTION
**Pull request by sliptonic**
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [ ] Branch rebased on latest master `git pull --rebase upstream master`
- [ ] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

### CODE REVIEW

File: src/Mod/Path/PathScripts/PathDeburr.py [hunk 1]
```diff
@@ -51,7 +51,14 @@ def translate(context, text, disambig=None):
...
 def toolDepthAndOffset(width, extraDepth, tool):
-    '''toolDepthAndOffset(width, extraDepth, tool) ... return tuple for given parameters.'''
+    '''toolDepthAndOffset(width, extraDepth, tool) ... return tuple for given\n
+       parameters.'''
+
+    if not (hasattr(tool, 'CuttingEdgeAngle')
+            and hasattr(tool, 'FlatRadius')
+            and hasattr(tool, 'Diameter')):
+        raise ValueError('Deburr requires tool with flatradius, diameter, and CuttingEdgeAngle\n')
+
```
Comment (mlampert) [inline src/Mod/Path/PathScripts/PathDeburr.py:61]: This prevents the use of endmills which was supported somewhat intentionally. Not sure if it's still needed now that there is profiling along edges - haven't used that though.

File: src/Mod/Path/PathScripts/PathDeburr.py [hunk 2]
```diff
@@ -30,7 +30,7 @@
 import PathScripts.PathOpTools as PathOpTools
 import math
 
-from PySide import QtCore
+from PySide import QtCore, QtGui
```
Comment (mlampert) [inline src/Mod/Path/PathScripts/PathDeburr.py:33]: Including `QtGui` in the op logic is kinda a no-no. This belongs into the ...Gui.py but not here.

File: src/Mod/Path/PathScripts/PathDeburr.py [hunk 3]
```diff
@@ -62,40 +69,52 @@ def toolDepthAndOffset(width, extraDepth, tool):
...
         obj.setEditorMode('Join', 2)  # hide for now
 
     def opExecute(self, obj):
         PathLog.track(obj.Label)
-        (depth, offset) = toolDepthAndOffset(obj.Width.Value, obj.ExtraDepth.Value, self.tool)
+        try:
+            (depth, offset) = toolDepthAndOffset(obj.Width.Value, obj.ExtraDepth.Value, self.tool)
+        except ValueError as e:
+            msg = "{} \n No path will be generated".format(e)
+            QtGui.QMessageBox.information(None, "Tool Error", msg)
```
Comment (mlampert) [inline src/Mod/Path/PathScripts/PathDeburr.py:111]: This makes it impossible to use the op in headless mode, which is the major reason we split all ops into op logic and ...Gui

File: src/Mod/Path/PathScripts/PathDeburrGui.py [hunk 4]
```diff
@@ -44,19 +44,39 @@
...
+        return super(TaskPanelBaseGeometryPage, self)
+
+    def addBaseGeometry(self, selection):
+        for sel in selection:
+            if sel.HasSubObjects:
+                # selectively add some elements of the drawing to the Base
+                for sub in sel.SubObjects:
+                    if isinstance(sub, Part.Face):
+                        if sub.normalAt(0, 0) != FreeCAD.Vector(0, 0, 1):
+                            PathLog.info(translate("Path", "Ignoring non-vertical Face"))
```
Comment (mlampert) [inline src/Mod/Path/PathScripts/PathDeburrGui.py:65]: "non-horizontal" ?

File: src/Mod/Path/PathScripts/PathDeburrGui.py [hunk 5]
```diff
@@ -109,6 +129,11 @@ def registerSignalHandlers(self, obj):
         self.form.value_W.editingFinished.connect(self.updateWidth)
         self.form.value_h.editingFinished.connect(self.updateExtraDepth)
 
+    def taskPanelBaseGeometryPage(self, obj, features):
+        '''taskPanelBaseGeometryPage(obj, features) ... return page for adding base geometries.'''
+        print(features)
```
Comment (mlampert) [inline src/Mod/Path/PathScripts/PathDeburrGui.py:134]: did you mean to remove this?

**Review by mlampert — CHANGES_REQUESTED** [review]

File: src/Mod/Path/PathScripts/PathDeburr.py (hunk 1 above)
Comment (sliptonic) [reply src/Mod/Path/PathScripts/PathDeburr.py:61]: This fix was just to catch the error that was painting the console red.  Has an earlier patch eliminated the user of endmills?

File: src/Mod/Path/PathScripts/PathDeburr.py (hunk 3 above)
Comment (sliptonic) [reply src/Mod/Path/PathScripts/PathDeburr.py:111]: confirmed

File: src/Mod/Path/PathScripts/PathDeburrGui.py (hunk 4 above)
Comment (sliptonic) [reply src/Mod/Path/PathScripts/PathDeburrGui.py:65]: zactly!

File: src/Mod/Path/PathScripts/PathDeburr.py (hunk 1 above)
Comment (sliptonic) [reply src/Mod/Path/PathScripts/PathDeburr.py:59]: Reworked the attribute checking and default to allow toolbit endmills as well as legacy endmills

File: src/Mod/Path/PathScripts/PathDeburr.py (hunk 2 above)
Comment (sliptonic) [reply src/Mod/Path/PathScripts/PathDeburr.py]: removed gui dependency.  Still using console messages.  Would be nice to throw the error and catch in the GUI but not sure how to do that if the taskpanel isn't open.

File: src/Mod/Path/PathScripts/PathDeburr.py (hunk 3 above)
Comment (sliptonic) [reply src/Mod/Path/PathScripts/PathDeburr.py]: resolved

File: src/Mod/Path/PathScripts/PathDeburrGui.py (hunk 5 above)
Comment (sliptonic) [reply src/Mod/Path/PathScripts/PathDeburrGui.py]: removed

**Comment by sliptonic** [comment]
I also made significant changes to correct the CW/CCW path generation.  I believe it is correct now but additional testing would be good for both deburr and engrave

**Review by mlampert — APPROVED** [review]

**Pull request merged**